chef/cookbooks/ceph/providers/client.rb

```diff
@@ -61,23 +61,30 @@ def get_caps(ceph_conf, admin_keyring, keyname)
 end
 
 def auth_set_key(ceph_conf, admin_keyring, keyname, caps)
+  raise "No caps defined for ceph_client resource!" if caps.empty?
   # try to add the key
   caps_str = caps.map { |k, v| "#{k} '#{v}'" }.join(" ")
   cmd = "ceph -k #{admin_keyring} -c #{ceph_conf} auth get-or-create #{keyname} #{caps_str}"
   get_or_create = Mixlib::ShellOut.new(cmd)
   get_or_create.run_command
-  unless get_or_create.stderr.scan(/EINVAL.*but cap.*does not match/).empty?
-    Chef::Log.info("Deleting old key with incorrect caps")
-    # don't remove custom existing caps for osd pools
-    caps_osd = get_caps(ceph_conf, admin_keyring, keyname)["osd"] + "," + caps["osd"]
-    caps["osd"] = caps_osd.split(",").collect(&:strip).uniq.join(",")
-    caps_str = caps.map { |k, v| "#{k} '#{v}'" }.join(" ")
-    cmd = "ceph -k #{admin_keyring} -c #{ceph_conf} auth get-or-create #{keyname} #{caps_str}"
-    # delete an old key if it exists and is wrong
-    Mixlib::ShellOut.new("ceph -k #{admin_keyring} -c #{ceph_conf} auth del #{keyname}").run_command
-    # try to create again
-    get_or_create = Mixlib::ShellOut.new(cmd)
-    get_or_create.run_command
+  return get_or_create.error! unless get_or_create.stderr.scan(/EINVAL.*but cap.*does not match/)
+  Chef::Log.info("Updating incorrect caps for #{keyname}")
+  # create list of uniq pools
+  pools = caps["osd"].split(",").collect(&:strip).map { |x| x.gsub(/allow [rwx]{1,3} /, "") }.uniq
+  # get current existing capabilities
+  cur_caps = get_caps(ceph_conf, admin_keyring, keyname)
+  if cur_caps["osd"] && !cur_caps["osd"].empty?
+    cur_caps["osd"].split(",").collect(&:strip).sort.uniq.each do |cap|
```

Comment:
Right, but the reason sort is there becasue the result needs to be "stable" independent of arbitrary dict reorderings between different runs. theoretically .uniq.sort would be slightly more efficient but I guess it does not matter whatsoever.